[WIP] odo NodeJS in a disconnected env. (#31914)
openshift/openshift-docs: modules/developer-cli-odo-creating-a-java-application-by-using-a-devfile-in-a-disconnected-cluster.adoc

  - name: springbootproject
    git:
      remotes:
        origin: "https://github.com/odo-devfiles/springboot-ex.git"
components:
  - name: tools
    container:
      image: quay.io/eclipse/che-java11-maven:nightly
      memoryLimit: 512Mi
      mountSources: true
      endpoints:
        - name: 'http-8080'
          targetPort: 8080
      volumeMounts:
        - name: m2
          path: /home/user/.m2
  - name: m2
    volume: {}
commands:
  - id: mvn-package
    exec:
      component: tools
      commandLine: "mvn -Dmaven.repo.local=/home/user/.m2/repository -Dhttp.proxyHost=<proxy-host> -Dhttp.proxyPort=<proxy-port> -Dhttps.proxyHost=<proxy-host> -Dhttps.proxyPort=<proxy-port> package"
      group:
        kind: build
        isDefault: true
  - id: run 
    exec:
      component: tools
      commandLine: "java -jar target/*.jar"
      group:
        kind: run
        isDefault: true
  - id: debug
    exec:
      component: tools
      commandLine: "java -Xdebug -Xrunjdwp:server=y,transport=dt_socket,address=${DEBUG_PORT},suspend=n -jar target/*.jar"
      group:
        kind: debug
        isDefault: true
----

. Create a Java application:
+
[source,terminal]
----

> $ odo create --devfile <path-to-your-devfile> --starter

Reviewer: I think it should be `odo create <cmp-name> --starter`, no?

Reviewer: Even though users provide their own similar devfile (which has starter project) they need to pass component name, I think

Author: @prietyc123 thank you! I've updated the commands and also added the warnings about the external dependencies (hopefully, the support will allow these). PTAL and let me know if you're okay with the changes.

Reviewer: @boczkowska Thanks for the changes 🙂 
LGTM

// Module included in the following assemblies:
//
// * cli_reference/developer_cli_odo/

[id="developer-cli-odo-creating-a-java-application-by-using-a-devfile-in-a-disconnected-cluster_{context}"]

= Creating a Java application by using a devfile in a disconnected cluster

[WARNING]
====

> This procedure is using external dependencies such as `quay.io/eclipse/che-java11-maven:nightly` or an example application `springboot-ex` that are not maintained by Red Hat. These dependencies are not maintained with the documentation and their functionality cannot be guaranteed. 

Reviewer 2: ```suggestion
This procedure uses external dependencies, such as `quay.io/eclipse/che-java11-maven:nightly` or an example application `springboot-ex`, that are not maintained by Red Hat and whose functionality may change. 
```
- s/is using/uses/ because the present progressive, "is using", suggests a temporary activity.
- We should not imply that the functionality of maintained software is guaranteed. Doing so has legal implications, which is outside the scope of our product documentation.

====

.Prerequisites
* You have created and logged into a disconnected cluster.

> * You have added `quay.io`, `registry.access.redhat.com`, `apache.org`, `quayio-production-s3.s3.amazonaws.com` URLs in your proxy. 

Reviewer 2: s/in your proxy/to your proxy configuration/ or something similar.

[...]
Downloaded from central: https://repo.maven.apache.org/maven2/org/codehaus/plexus/plexus-utils/3.2.1/plexus-utils-3.2.1.jar (262 kB at 813 kB/s)
[INFO] Replacing main artifact with repackaged archive
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  19.638 s
[INFO] Finished at: 2021-02-24T08:59:30Z
[INFO] ------------------------------------------------------------------------
 ✓  Executing mvn-package command "mvn -Dmaven.repo.local=/home/user/.m2/repository -Dhttp.proxyHost=<proxy-host> -Dhttp.proxyPort=<proxy-port> -Dhttps.proxyHost=<proxy-host> -Dhttps.proxyPort=<proxy-port> package" [23s]
 •  Executing run command "java -jar target/*.jar"  ...
I0224 14:29:30.557676   34426 exec.go:27] Executing command [/opt/odo/bin/supervisord ctl start devrun] for pod: java-maven-5b8f99fcdb-9dnk6 in container: tools
devrun: started
 ✓  Executing run command "java -jar target/*.jar" [3s]

Pushing devfile component java-maven
 ✓  Changes successfully pushed to component
----

> . Verify that the application has stared by displaying logs:

Reviewer 2: ```suggestion
. Verify that the application has started by displaying logs:
```

Author: @rolfedh what I meant to say here was "Display the logs to verify that the application has started" - WDYT of this wording?

Reviewer 2: That's even better.
